core/src/prediction/LLCausalPredictionStrategy.cpp

```diff
@@ -148,7 +148,149 @@ std::vector<double> LLCausalPredictionStrategy::compute_variance(
         const Data* train_data,
-        const Data* data,
+        const Data* test_data,
         size_t ci_group_size){
-  return { 0.0 };
+
+  double lambda = lambdas[0];
+
+  size_t num_variables = linear_correction_variables.size();
+  size_t num_nonzero_weights = weights_by_sampleID.size();
+
+  std::vector<size_t> sample_index_map(train_data->get_num_rows());
+  std::vector<size_t> indices(num_nonzero_weights);
+
+  Eigen::MatrixXd weights_vec = Eigen::VectorXd::Zero(num_nonzero_weights);
+  {
+    size_t i = 0;
+    for (auto& it : weights_by_sampleID) {
+      size_t index = it.first;
+      double weight = it.second;
+      indices[i] = index;
+      sample_index_map[index] = i;
+      weights_vec(i) = weight;
+      i++;
+    }
+  }
+
+  size_t dim_X = 2 * num_variables + 2;
+  Eigen::MatrixXd X (num_nonzero_weights, dim_X);
+  Eigen::MatrixXd Y (num_nonzero_weights, 1);
+  size_t treatment_index = num_variables + 1;
+
+  for (size_t i = 0; i < num_nonzero_weights; ++i) {
+    // Index of next neighbor with nonzero weights
+
+    size_t index = indices[i];
+    double treatment = train_data->get_treatment(index);
+
+    // Intercept
+    X(i, 0) = 1;
+
+    // Regressors
+    for (size_t j = 0; j < num_variables; ++j){
+      size_t current_predictor = linear_correction_variables[j];
+      // X - x0 column
+      X(i,j+1) = train_data->get(index, current_predictor) -
+                 test_data->get(sampleID, current_predictor);
+
+      // (X - x0)*W column
+      X(i, treatment_index + j + 1) = X(i, j+1) * treatment;
+    }
+
+    // Treatment (just copied)
+    X(i, treatment_index) = treatment;
+
+    // Outcome (just copied)
+    Y(i) = train_data->get_outcome(index);
+  }
+
+  // find ridge regression predictions
+  Eigen::MatrixXd M_unpenalized (dim_X, dim_X);
+  M_unpenalized.noalias() = X.transpose() * weights_vec.asDiagonal() * X;
+
+  Eigen::MatrixXd M;
+  M = M_unpenalized;
+  if (!weight_penalty) {
+    double normalization = M_unpenalized.trace() / dim_X;
+
+    // standard ridge penalty
+    for(size_t j = 1; j < dim_X; ++j){
+      if(j != treatment_index){
+        M(j, j) += lambda * normalization;
+      }
+    }
+  } else {
+    // covariance ridge penalty
+    for(size_t j = 1; j < dim_X; ++j){
+      if(j != treatment_index){
+        M(j, j) += lambda * M(j, j);
+      }
+    }
+  }
+
+  Eigen::VectorXd theta = M.ldlt().solve(X.transpose()*weights_vec.asDiagonal()*Y);
+
+  Eigen::VectorXd e_trt = Eigen::VectorXd::Zero(dim_X);
+  e_trt(treatment_index) = 1.0;
+  Eigen::VectorXd zeta = M.ldlt().solve(e_trt);
+
+  Eigen::VectorXd X_times_zeta = X * zeta;
+  Eigen::VectorXd local_prediction = X * theta;
+  Eigen::VectorXd pseudo_residual = Eigen::VectorXd::Zero(num_nonzero_weights);
+
+  for (size_t i = 0; i < num_nonzero_weights; i++) {
+    pseudo_residual(i) = X_times_zeta(i) * (Y(i) - local_prediction(i,1) - X(i, treatment_index)*local_prediction(i, treatment_index));
```

Comment:
The local linear regression adjustment appears to be missing here?

Reply:
I may be thinking through the math incorrectly; but I think it's analogous to the pseudo-residuals for LLF, which are:

pseudo_residual(i) = X_times_zeta(i) * (Y(i) - local_prediction(i));

Here, we just de-mean Y(i) by the local_prediction(0) term - the 1 is a typo I'll fix, forgot about C++ indexing- and then subtract out (W-W.hat)*tau.hat.local. The linear correction goes into our prediction for tau.hat.local, but I think the pseudo-residual should just have the actual prediction.

Reply:
@swager let me know what you think! All other comment responses are now in :)

Reply:
Ok I see -- now I understand what's up here. Just to make sure I'm not missing anything, though, could you please write up a short tex document that spells out the algebra going on here (including the dimension of all the objects)? In particular, what is the dimension of "local_prediction"?

Reply:
Ok good check - this code was actually a bit more complicated than it should have been! Doc is reticent to attach, but I'll email it to you and try to add it to this conversation for future reference.

Reply:
Got it, thanks! I had been looking at the outdated version of the code above -- that's why I got confused.